Alternate solution for timeout blur

bartpeeters opened the pull request:
Follow up from https://github.com/gpbl/react-day-picker/issues/579#issuecomment-353798385, should replace #525.

---

Issue:
1. Open Firefox (its not an issue in chrome) https://codesandbox.io/s/XDAE3x0W8
2. Click on day picker input (it shows calendar)
3. Click on navigation bar element (next, prev or nav title elements)
4. Main day picker input loses own focus and calendar stays open
5. Now when you click anywhere outside of calendar. Calendar is still open but is should close.

The fix using a setTimeout onBlur to call this.input.focus() caused new issues: 
- A flicker in Chrome everytime you click inside the daypicker
- When using a custom captionElement containing a <select> it would not open because the timeout focus cancelled it

This alternate solution is to remove the timeout and instead add an onBlur on the overlay doing the same thing as the onBlur of the input.

gpbl commented:
Thanks @bartpeeters for your PR. I want to try it first, sure we have to remove that timeout thing since it doesn't well. 

I don't understand why the new `span` element should fire the `onBlur` event [here](https://github.com/gpbl/react-day-picker/pull/598/files#diff-e2a925e665d836a09575f79095191746R415). It shouldn't ever get the focus. Have you find the reason behind it?

bartpeeters commented:
@gpbl tabIndex on the daypicker ([here](https://github.com/gpbl/react-day-picker/blob/8c69e20419bfc3a601de7ee267c2436707f3ed62/src/DayPicker.js#L553)) makes it so the span is focusable.

gpbl commented on src/DayPickerInput.js:
```diff
@@ -417,22 +412,24 @@ export default class DayPickerInput extends React.Component {
-          ref={el => (this.daypicker = el)}
-          onTodayButtonClick={onTodayButtonClick}
-          {...dayPickerProps}
+      <span onBlur={this.handleInputBlur}>
```
My suspect is that this line does nothing. We should add an unit test to it.

gpbl commented on test/daypickerinput/events.js:
```diff
@@ -52,18 +52,6 @@ describe('DayPickerInput', () => {
         wrapper.find('input').simulate('blur');
         expect(onBlur).toHaveBeenCalledTimes(1);
       });
-      it('should focus the input if blur after clicking the overlay', done => {
```
This test must stay here :) Why does your change make it fail?

bartpeeters commented:
@gpbl Thanks for the feedback.
I implemented another solution, using the onFocus event of the overlay to always make sure the DaypickerInput input field is focused. 

But do we want this? What if we have a custom overlay that includes an input? Now this input is unusable because it can't be focused. We can check if the nodeName of the focus event target is an input, but I don't really know if that is the way to go.

I also replaced the clickInside with the timeout by using a ref on the Overlay and checking if the relatedTarget of blur event is inside the overlay. I'm not sure that this is better than using a timeout.

About the bubbling of the onFocus and onBlur on the span which contains the overlay, the React Synthetic onFocus and onBlur events do bubble, unlike the normal blur and focus events.

gpbl commented:
@bartpeeters thanks for the updates. I've not checked your changes yet, however I'm proposing a different solution:

- reset this line to the original behavior, which was working well except on Firefox:

```diff
if (this.clickedInside) { 
-    this.blurTimeout = setTimeout(() => this.input.focus(), 0); 
+   this.input.focus(); 
}
```

- on Firefox, we just disable this to resolve https://github.com/gpbl/react-day-picker/issues/579:

```diff
- if (this.clickedInside) { 
+ if (!isFirefox && this.clickedInside) { 
```

- for the case when the calendar has other focusable elements inside, such as selects or input fields (as in https://github.com/gpbl/react-day-picker/issues/599), we allow the implementer to disable this focus with a new DayPickerInput prop: `keepFocus` that can be set to `false`.

bartpeeters commented:
@gpbl I like the idea of a new prop `keepFocus`.

The solution you propose has the same effect as these changes except they dont use `isFirefox`, so maybe you can take a look at the code. I'm also up to add the `keepFocus` prop and some tests if you approve of these changes.

mattfysh commented:
@gpbl how can I help push this forward? would really appreciate being able to set `keepFocus` to `false`

bartpeeters commented:
@mattfysh I just added the `keepFocus` prop and added api documentation for it. I think now we just need to wait till @gpbl has some time to review these changes.

ghaynes1292 commented on docs/src/pages/api/DayPickerInput.js:
```diff
@@ -246,6 +247,16 @@ function MyDayPickerInput(props) {
       </p>
+      <h3>
+        <Anchor id="keepFocus" />
+        keeFocus <code>boolean = true</code>
```
keepFocus

bartpeeters replied:
Thanks!

rdiwelm commented:
Hey guys! How soon can we expect a new version release with the new `keepFocus` prop?

gpbl commented:
@rdiwelm soon ! I have been pretty busy the last month, hoping to catch up all this weekend 🙏🏽 sorry for the lack of communication.

roginfarrer commented:
I also have a feature depending on this release!

rdiwelm commented:
Hey guys! Sorry to be asking again. Is there any ETA on this? We're kinda blocked too. :/

kradical commented:
@roginfarrer @rdiwelm I was blocked by this too so we forked this dependency until this change gets merged. I understand that is not always a viable option.

gpbl commented:
Published as [v7.1.0](https://www.npmjs.com/package/react-day-picker).

rdiwelm commented:
Awesome! Thanks a lot, guys! :)
